### ISSUE TITLE
ogr2ogr with MySQL 8: POINT column doesn't have SRID attribute

### ISSUE THREAD

**Issue by YoshiakiYamasaki** [OSGeo/gdal#1015]
I imported shape file using ogr2ogr and I checked created table. Created table doesn't have SIRD attribute on POINT column.

## Expected behavior and actual behavior.
Created table have SIRD attribute on POINT column.

Note: 
POINT column of MySQL 8.0 can have SRID attribute. I think adding SRID attribute is desirable to avoid mistakes.

MySQL 8.0 Reference Manual / 11.5.1 Spatial Data Types
https://dev.mysql.com/doc/refman/8.0/en/spatial-type-overview.html
=> Columns with a spatial data type can have an SRID attribute, to explicitly indicate the spatial reference system (SRS) for values stored in the column.

## Steps to reproduce the problem.
1. Import shape file using ogr2ogr
ex) C:\> ogr2ogr -f "MySQL" MYSQL:geotest,user=root,password=root C:\work\point_2449.shp

2. Execute "show create table" for created table
ex) mysql> show create table point_2449\G
*************************** 1. row ***************************
       Table: point_2449
Create Table: CREATE TABLE `point_2449` (
  `OGR_FID` int(11) NOT NULL AUTO_INCREMENT,
  `SHAPE` geometry NOT NULL,
  `gid` decimal(10,0) DEFAULT NULL,
  `point_name` varchar(254) DEFAULT NULL,
  UNIQUE KEY `OGR_FID` (`OGR_FID`),
  SPATIAL KEY `SHAPE` (`SHAPE`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set (0.00 sec)

* I hope modify CREATE TABLE as follows.
Before:  `SHAPE` geometry NOT NULL,
After:  `SHAPE` geometry NOT NULL SRID 2449,

## Operating system
Windows 10 x64

## GDAL version and provenance
C:\> ogr2ogr --version
GDAL 2.3.2, released 2018/09/21

**Comment by YoshiakiYamasaki** [comment]
I found an important problem regarding this issue. **MySQL can not use a spatial index if no SRID is defined for the column**. Therefore, I hope that the importance of this issue goes up.

[MySQL Document]
MySQL 8.0 Reference Manual / 8.3.3 SPATIAL Index Optimization
https://dev.mysql.com/doc/refman/8.0/en/spatial-index-optimization.html

> The optimizer considers SPATIAL indexes only for SRID-restricted columns:
>    Indexes on columns restricted to a Cartesian SRID enable Cartesian bounding box computations.
>    Indexes on columns restricted to a geographic SRID enable geographic bounding box computations. 
> 
> **The optimizer ignores SPATIAL indexes on columns that have no SRID attribute** (and thus are not SRID-restricted).

**Comment by rouault** [comment]
> I hope that the importance of this issue goes up

This is free and open source software. See http://nyalldawson.net/2016/08/how-to-effectively-get-things-changed-in-qgis/ that applies to GDAL as well

**Comment by YoshiakiYamasaki** [comment]
I see. Thank you for the information. I'll try to submit pull request later.

To be honest, this is first time submitting pull request. Therefore, maybe it needs some time. However, I want to try it!

**Comment by miurahr** [comment]
@YoshiakiYamasaki  How it should do when unsupported SRID is used in source data?

MySQL 8.0 has a limitation on SRID hint; 
https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/spatial-type-overview.html
>  InnoDB tables permit SRID values for Cartesian and geographic SRSs. MyISAM tables permit SRID values for Cartesian SRSs.

Geographic SRS is such as 4326, and Cartesian is such as 27700.

**Comment by YoshiakiYamasaki** [comment]
> @YoshiakiYamasaki How it should do when unsupported SRID is used in source data?
> 
> MySQL 8.0 has a limitation on SRID hint;
> https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/spatial-type-overview.html
> 
> > InnoDB tables permit SRID values for Cartesian and geographic SRSs. MyISAM tables permit SRID values for Cartesian SRSs.
> 
> Geographic SRS is such as 4326, and Cartesian is such as 27700.

In that case, the following error is returned.
> ERROR 1178 (42000): The storage engine for the table doesn't support geographic spatial reference systems

If you specify an unsupported SRID, the following error is returned.
> ERROR 3548 (SR001): There's no spatial reference system with SRID 277000.

------------------------------
[Test log]
* Default Storage Engine is InnoDB.

mysql> CREATE TABLE test_InnoDB(id INT, geo GEOMETRY SRID 4236);
Query OK, 0 rows affected (0.15 sec)

mysql> drop table test_InnoDB;
Query OK, 0 rows affected (0.10 sec)

mysql> CREATE TABLE test_InnoDB(id INT, geo GEOMETRY SRID 27700);
Query OK, 0 rows affected (0.04 sec)

mysql> drop table test_InnoDB;
Query OK, 0 rows affected (0.07 sec)

mysql> CREATE TABLE test_MyISAM(id INT, geo GEOMETRY SRID 4236) ENGINE=MyISAM;
ERROR 1178 (42000): The storage engine for the table doesn't support geographic spatial reference systems
mysql> CREATE TABLE test_MyISAM(id INT, geo GEOMETRY SRID 27700) ENGINE=MyISAM;
Query OK, 0 rows affected (0.02 sec)

mysql> drop table test_MyISAM;
Query OK, 0 rows affected (0.02 sec)

mysql> CREATE TABLE test_InnoDB(id INT, geo GEOMETRY SRID 277000);
ERROR 3548 (SR001): There's no spatial reference system with SRID 277000.
------------------------------

**Comment by miurahr** [comment]
@YoshiakiYamasaki What is a correct behavior for  GDAL/ogr when accepting such SRID value?

   Simply adding SRID block brings  autotest case `test_ogr_mysql_longlat` failure.

```
     def test_ogr_mysql_longlat():
    
        if gdaltest.mysql_ds is None:
            pytest.skip()
    
        srs = osr.SpatialReference()
        srs.SetFromUserInput('WGS84')
        lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_longlat',
                                            geom_type=ogr.wkbPoint,
                                            srs=srs,
                                            options=['ENGINE=MyISAM'])
>       f = ogr.Feature(lyr.GetLayerDefn())
E       AttributeError: 'NoneType' object has no attribute 'GetLayerDefn'

ogr_mysql.py:880: AttributeError
-------------------------------------------- Captured stderr call ---------------------------------------------
MYSQL: Creating layer ogr_mysql_longlat.
MYSQL: Geometry Column Name SHAPE.
MYSQL: FID Column Name OGR_FID.
ERROR 1: MySQL error message:The storage engine for the table doesn't support geographic spatial reference systems Description: CREATE TABLE `ogr_mysql_longlat` (    OGR_FID INT UNIQUE NOT NULL AUTO_INCREMENT,    SHAPE GEOMETRY NOT NULL /*!80003 SRID 4326 */) ENGINE = MyISAM
```

Without SRID, before adding SRID, it passes.
Is it expected behavior for MySQL users? Should the case  fail because of mis-configured combination of ENGINE and SRID?

**Comment by miurahr** [comment]
> This also means that the geometry_columns table is no longer needed since MySQL 8.0 uses the INFORMATION_SCHEMA.ST_GEOMETRY_COLUMNS table.

It seems the table is necessary when using older version of MySQL and MariaDB. So the driver should handle two different way according to server version.

**Comment by miurahr** [comment]
>This also means that the geometry_columns table is no longer needed since MySQL 8.0 uses the INFORMATION_SCHEMA.ST_GEOMETRY_COLUMNS table.
@joshbernfeld I'd like to update it as another PR than #2111 for review because it affect many part of mysql driver.

**Comment by YoshiakiYamasaki** [comment]
> 
> 
> @YoshiakiYamasaki What is a correct behavior for GDAL/ogr when accepting such SRID value?
> 
> Simply adding SRID block brings autotest case `test_ogr_mysql_longlat` failure.
> 
> ```
>      def test_ogr_mysql_longlat():
>     
>         if gdaltest.mysql_ds is None:
>             pytest.skip()
>     
>         srs = osr.SpatialReference()
>         srs.SetFromUserInput('WGS84')
>         lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_longlat',
>                                             geom_type=ogr.wkbPoint,
>                                             srs=srs,
>                                             options=['ENGINE=MyISAM'])
> >       f = ogr.Feature(lyr.GetLayerDefn())
> E       AttributeError: 'NoneType' object has no attribute 'GetLayerDefn'
> 
> ogr_mysql.py:880: AttributeError
> -------------------------------------------- Captured stderr call ---------------------------------------------
> MYSQL: Creating layer ogr_mysql_longlat.
> MYSQL: Geometry Column Name SHAPE.
> MYSQL: FID Column Name OGR_FID.
> ERROR 1: MySQL error message:The storage engine for the table doesn't support geographic spatial reference systems Description: CREATE TABLE `ogr_mysql_longlat` (    OGR_FID INT UNIQUE NOT NULL AUTO_INCREMENT,    SHAPE GEOMETRY NOT NULL /*!80003 SRID 4326 */) ENGINE = MyISAM
> ```
> 
> Without SRID, before adding SRID, it passes.
> Is it expected behavior for MySQL users? Should the case fail because of mis-configured combination of ENGINE and SRID?

@miurahr 
I think returning the MySQL error directly is the correct behavior.

However, in that case, we can't import data into MySQL, so it is desirable to provide the following workaround:.

  1. Options that can be create SQL statement to a text file only(ogr2ogr does not execute SQL directly). I've submitted the following feature request in the past. If this option is provided, MySQL users can fix the problem themselves by modifying the SQL statement.
https://github.com/OSGeo/gdal/issues/1090

  2. Options that can be imported without an SRID, such as -no-grid, are provided as options for ogr2ogr (Same behavior as current behavior)

BTW, there are following background. Today, few users want to use GIS feature with MyISAM. Almost users want to use GIS feature with InnoDB.

[Note]
- MyISAM is rarely used now. (Beginning with MySQL 5.5, the default storage engine has been changed to InnoDB.)
- In MySQL 5.7, GIS features have been reimplemented, with InnoDB as the premise.

**Comment by miurahr** [comment]
'ENGINE=MyISAM' option in these test code  were added in 7 years ago in days of MySQL 5.5.
https://github.com/OSGeo/gdal/commit/7ef89c6062281857071a2e39e9e37e7fc9e21f97

It would be OK to remove  it now.

Command 'ogr2ogr' use default engine that is InnoDB, so problems may not be happen.
Developers can specify an 'ENGINE=MyISAM' option to API when  using MyISAM .

**Comment by YoshiakiYamasaki** [comment]
I agree. I think there is no problem to delete 'ENGINE=MyISAM' option in the test code.

**Comment by miurahr** [comment]
Is it safe to remove ENGINE=MyISAM for testing on MariaDB? 
Because GDAL also supports MariaDB with the same driver.

**Comment by YoshiakiYamasaki** [comment]
I think it is safe.
 - The InnoDB storage engine is now the default for MariaDB.
 - The MariaDB page also says that "MyISAM is MySQL's oldest storage engine. There is usually little reason to use it except for legacy purposes.".
    https://mariadb.com/kb/en/library/choosing-the-right-storage-engine/

**Comment by miurahr** [comment]
@YoshiakiYamasaki @joshbernfeld @rouault I believe #2111 is ready for review.
And I'd like to start a topic branch to use the INFORMATION_SCHEMA.ST_GEOMETRY_COLUMNS table as another PR.

**Comment by miurahr** [comment]
@joshbernfeld could you give a feedback for https://github.com/OSGeo/gdal/pull/2130 that uses the INFORMATION_SCHEMA.ST_GEOMETRY_COLUMNS table.?

**Comment by joshbernfeld** [comment]
@miurahr Sure, will review tomorrow.

**Comment by YoshiakiYamasaki** [comment]
@miurahr @joshbernfeld @rouault Thank you very much!!
